lms/djangoapps/course_api/api.py

```diff
@@ -170,19 +172,38 @@ def list_course_keys(request, username, role):
     """
     user = get_effective_user(request.user, username)
 
-    course_keys = CourseOverview.get_all_course_keys()
+    all_course_keys = CourseOverview.get_all_course_keys()
 
     # Global staff have access to all courses. Filter courses for non-global staff.
     if GlobalStaff().has_user(user):
-        return course_keys
-
-    return LazySequence(
-        (
-            course_key for course_key in course_keys
-            if has_access(user, role, course_key)
-        ),
-        est_len=len(course_keys)
-    )
+        return all_course_keys
+
+    if role == 'staff':
+        # This short-circuit implementation bypasses has_access() which we think is too slow for some users when
+        # evaluating staff-level course access for Insights.  Various tickets have context on this issue: CR-2487,
+        # TNL-7448, DESUPPORT-416, and probably more.
+        #
+        # This is a simplified implementation that does not consider org-level access grants (e.g. when course_id is
+        # blank).
+        filtered_course_keys = (
+            CourseAccessRole.objects.filter(
+                user=user,
+                # Having the instructor role implies staff access. This fact was reverse-engineered from unit tests.
+                role__in=['staff', 'instructor'],
```

Comment:
For reviewers: This is the line I changed.  Now we include "instructor" as implied staff.

Reply:
Does that mean we need to de-dupe the course_ids?

Also, since we're supporting paging, I'm guessing there should be an order_by?

Reply:
So something like this, maybe?

```python
            CourseAccessRole.objects.filter(
                user=user,
                # Having the instructor role implies staff access. This fact was reverse-engineered from unit tests.
                role__in=['staff', 'instructor'],
            )
            .exclude(course_id=CourseKeyField.Empty)
            .order_by('course_id')
            .distinct('course_id')
            .values_list('course_id', flat=True)
```

Reply:
Oh wait, `distinct` only accepts args with Postgres, one sec...

Reply:
I think this would work?

```python
CourseAccessRole.objects.filter(
    user=user,
    # Having the instructor role implies staff access. This fact was reverse-engineered from unit tests.
    role__in=['staff', 'instructor'],
)
.exclude(course_id=CourseKeyField.Empty)
.order_by('course_id')
.values_list('course_id', flat=True)
.distinct()
```

Reply:
Regardless, I did just make this change and add a unit test for it.